File: Sorting_Searching/binary_search.cpp

#include <iostream>
using namespace std;

// Binary Search: Iterative
int binarySearch(int arr[], int N, int ele)
{
	int low = 0;
	int high = N - 1;
	while (low <= high)
	{
		int mid = low + (high - low) / 2;
		if (arr[mid] == ele)
			return mid;
		if (arr[mid] > ele)
			high = mid - 1;
		else
			low = mid + 1;
	}
	return -1;
}

/* Binary Search: Recursive
int binarySearch(int arr[], int N, int ele, int low, int high)
{
	if (low <= high)
	{
		int mid = low + (high - low) / 2;
		if (arr[mid] == ele)
			return mid;
		if (arr[mid] > ele)
			binarySearch(arr, N, ele, low, mid - 1);
		binarySearch(arr, N, ele, mid + 1, high);
	}
	return -1;
} */

// Main function
int main()
{
	int *array, N, ele;
	cout << "\nEnter the size of array: ";
	cin >> N;

	array = new int[N];
	cout << "Enter the elements of sorted array: ";
	for (int i = 0; i < N; i++)
		cin >> array[i];

	cout << "Enter an element to be searched: ";
	cin >> ele;

	int res = binarySearch(array, N, ele);
	(res == -1) ? cout << "\nResult: -1 i.e. Element not found!" : cout << "\nElement found at index " << res << "!";

	cout << endl;
	return 0;
}

File: arrays/Searching/BinarySearch.java
package arrays.Searching;

import java.util.Scanner;

public class BinarySearch {
    Scanner sc = new Scanner(System.in);

    private int[] readArray(){
        int size = sc.nextInt();
        int[] array = new int[size];
        for(int i = 0 ; i < array.length; i++){
            array[i] = sc.nextInt();
        }
        return array;
    }

    private int binarySearch(int[] array, int target){
        int low = 0;
        int high = array.length - 1;
        while( low <= high){
            int mid = low + (high - low) / 2;
            if(target == array[mid]){
                return mid;
            }else if(target > array[mid]){
                low = mid + 1;
            }else{
                high = mid - 1;
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        BinarySearch search = new BinarySearch();
        int[] array = search.readArray();
        int target = search.sc.nextInt();
        System.out.println(target + " is Found At index " +
                search.binarySearch(array, target));

    }
}
